pkg/operator/connectivitycheckcontroller/connectivity_check_controller.go

```diff
@@ -0,0 +1,252 @@
+	}
+	return templates
+}
+
+// listAddressesForKubeAPIServerServiceEndpoints returns kas api service endpoints ip
+func listAddressesForKubeAPIServerServiceEndpoints(endpointsLister corev1listers.EndpointsLister, recorder events.Recorder) []string {
+	var results []string
+	endpoints, err := endpointsLister.Endpoints("openshift-kube-apiserver").Get("apiserver")
+	if err != nil {
+		recorder.Warningf("EndpointDetectionFailure", "unable to determine openshift-kube-apiserver apiserver endpoints: %v", err)
+		return nil
+	}
+	for _, subset := range endpoints.Subsets {
+		for _, address := range subset.Addresses {
+			for _, port := range subset.Ports {
+				results = append(results, net.JoinHostPort(address.IP, strconv.Itoa(int(port.Port))))
+			}
+		}
+	}
+	return results
+}
+
+func getTemplatesForStorageEndpoints(operatorSpec *operatorv1.OperatorSpec, recorder events.Recorder) []*v1alpha1.PodNetworkConnectivityCheck {
+	var templates []*v1alpha1.PodNetworkConnectivityCheck
+	for _, address := range listAddressesForStorageEndpoints(operatorSpec, recorder) {
+		templates = append(templates, newPodNetworkProductivityCheck("storage-endpoint", address, withTlsClientCert("etcd-client")))
+	}
+	return templates
+}
+
+func listAddressesForStorageEndpoints(operatorSpec *operatorv1.OperatorSpec, recorder events.Recorder) []string {
+	var results []string
+	var observedConfig map[string]interface{}
+	if err := yaml.Unmarshal(operatorSpec.ObservedConfig.Raw, &observedConfig); err != nil {
+		recorder.Warningf("EndpointDetectionFailure", "failed to unmarshal the observedConfig: %v", err)
+		return nil
+	}
+	urls, _, err := unstructured.NestedStringSlice(observedConfig, "storageConfig", "urls")
+	if err != nil {
+		recorder.Warningf("EndpointDetectionFailure", "couldn't get the storage config urls from observedConfig: %v", err)
+		return nil
+	}
+	for _, rawStorageConfigURL := range urls {
+		storageConfigURL, err := url.Parse(rawStorageConfigURL)
+		if err != nil {
+			recorder.Warningf("EndpointDetectionFailure", "couldn't parse a storage config url from observedConfig: %v", err)
+			continue
+		}
+		results = append(results, storageConfigURL.Host)
+	}
+	return results
+}
+
+var checkNameRegex = regexp.MustCompile(`[.:\[\]]+`)
+
+func newPodNetworkProductivityCheck(label, address string, options ...func(*v1alpha1.PodNetworkConnectivityCheck)) *v1alpha1.PodNetworkConnectivityCheck {
+	check := &v1alpha1.PodNetworkConnectivityCheck{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "$(SOURCE)-to-" + label + "-" + checkNameRegex.ReplaceAllLiteralString(address, "-"),
+			Namespace: operatorclient.TargetNamespace,
+		},
+		Spec: v1alpha1.PodNetworkConnectivityCheckSpec{
+			TargetEndpoint: address,
+		},
+	}
+	for _, option := range options {
+		option(check)
+	}
+	return check
+}
+
+func withTlsClientCert(secretName string) func(*v1alpha1.PodNetworkConnectivityCheck) {
```

Comment:
nit: we use TLS

Reply:
@deads2k  https://github.com/openshift/cluster-openshift-apiserver-operator/pull/368